### ISSUE TITLE
Support including submodules in their parent's page

### ISSUE THREAD

**Issue by lpw25** [ocaml/odoc#564]
Sometimes people have small sub-modules and they would prefer to get something like:

```ocaml
  module Foo : sig
  
    type t
    
    val x : t
    
  end
```
in their docs instead of:
```ocaml
  module Foo : sig ... end
```
and a separate page for `Foo`.

It would be nice to allow users to annotate modules with some kind of tag to indicate that they would like this behaviour. It would require support in the model since links that used to go to the separate `Foo` page would now need to go to its parent page.

**Comment by lpw25** [comment]
I thought we already had an issue for this, but I couldn't find it. I also think this feature request is pretty low priority at the moment.

**Comment by ulrikstrid** [comment]
Semi-related: https://github.com/ocaml/odoc/issues/323

Also, I really want this 😄

**Comment by Drup** [comment]
I was slightly confused so I checked: It's already supported everywhere .... except there are no annotations for this in the syntax, unlike `include`. The models and everything down already does the right thing. In fact, both the latex and man output automatically unfold submodules, and the links are still correct in latex.

**Comment by lpw25** [comment]
I don't see how the links can be correct. You need to add a new form to the path syntax to represent "project submodule 'Foo' that is in the same page as its parent" and then resolving has to detect an annotation on the definitions and decide to use that projection rather than the ordinary one. Otherwise it will try to link to a non-existent page for the submodule.

**Comment by lpw25** [comment]
To be clear, this is fine in latex because there is only one possible projection in latex. In the html there will be two possible projections and resolution needs to produce the correct one.

**Comment by ulrikstrid** [comment]
Related to this I would like it if we can expand `module type` in the file. My usecase is that I have a "API" `module type` that then different modules implement. So I would like to showcase the API in the main module so that it's immediately obvious instead of clicking into one of the implementations.

**Comment by jonludlam** [comment]
Right - we'll also need to be careful with anchors as doing it naively will cause name clashes.

**Comment by jonludlam** [comment]
I'll rename this so it applies to all expansions rather than just modules.

**Comment by dmbaturin** [comment]
I also would like to see this implemented.

**Comment by dbuenzli** [comment]
I think this is a pretty bad idea and should not be implemented. I'll just repeat here what I said [in that discussion](https://discuss.ocaml.org/t/inline-sub-module-documentation-with-odoc/8293) on discuss.

>  I'm not saying you won't be able to find examples where the feature makes sense. 
> 
> The problem is that everyone will have a different definition of what is small, or some will have the impression that it's always better to inline everything regardless of the size of the module. So in aggregate docsets like [those generated](https://b0-system.github.io/odig/doc/) by odig for your opam switches, you get an unpredictable structure for end users (starting with the hackability of URLs). 
>
> Now someone of course will say oh but you should be able to override it on the cli of `odoc` but that's not good either because the way you write your docs does depend on the structure of the output, so you don't want to be betrayed by it later.

**Comment by dmbaturin** [comment]
Do you think it would be feasible to implement it as a documentation comment directive?

**Comment by dbuenzli** [comment]
> Do you think it would be feasible to implement it as a documentation comment directive?

I don't see how this solves the problems I mentioned, namely that 1) Everyone is going to have it's own notion of what is small 2) If you allow to force ignoring the directive then you betray what the author had in mind. 

If you hadn't allowed it in the first place the author would maybe have worked around the inability to inline in pathological cases in a creative way (e.g. by authoring a list with direct links).

Another argument against it is that 3) What is small may become big later when the code evolves (or a new maintainer has a different notion of what in small). Then the author or new maintainer changes the inlining desire and you just broke external links into the docs.

It is important for `odoc` to enforce a regular structure on what OCaml documentation should be and look like. Both for end users so that they don't have to learn new conventions each time they peruse a new package and for authors so that they can have reasonable expectations on what it will look like in all the contexts (odig, project website, docs.ocaml.org) as this does influence the way you design your documents and linking structure.

I think it already became too flexible, if you compare odig/dune generated docsets with what is currently on v3.ocaml.org there is already one pretty bad difference which as an author I dislike. Namely it is wrong to have "in this package" before "on this page", on a laptop screen you don't see "on this page". Yet that's what you are most interested in when you get on a page since it gives you its signposts, but in most cases this will end up burried behind scrolling which is really bad.

I always take a lot of care to design my heading structure so that a good toc is generated for efficient access to elements on the page. And that work is totally destroyed by [this rendering](https://v3.ocaml.org/p/brr/0.0.2/doc/ffi_cookbook.html) on v3.ocaml.org (also for some reason the page is missing the title and preamble, see [here](https://erratique.ch/software/brr/doc/ffi_cookbook.html) for a good rendering).

**Comment by dmbaturin** [comment]
@github-actions No, bot, just because no one is actively working on it now, the issue doesn't become less relevant. ;)

**Comment by mjambon** [comment]
> It is important for `odoc` to enforce a regular structure on what OCaml documentation should be and look like.

I understand you would like consistency across docs from the OCaml ecosystem but I disagree that this is more valuable than giving us an option.

In the project I'm working on, the value of having two tiny submodules stay on the same page just like they were in the original mli file is greater for me and for my users than consistency with web pages for unrelated libraries.

Here's a PR that adds support for `@inline` in front of submodules: https://github.com/ocaml/odoc/pull/1408 (first shot generated by Claude Code; I don't expect you to merge it)

**Comment by dbuenzli** [comment]
> the value of having two tiny submodules stay on the same page just like they were in the original mli file is greater for me and for my users than consistency with web pages for unrelated libraries.

What is missing from your point is that your users are unlikely to use a single library. You may not value that consistency in this case because *you* are the one making the choice. But in reality your library is unlikely to be the center of the world they operate in.

The day you are inflicted a seemingly random grab bag of inlines depending on the API docs you peruse because every API author was allowed to decide the matter based on their taste, your opinions on the matter could perhaps change.

**Comment by dbuenzli** [comment]
Also I should add if the enclosing module is only here for namespacing purposes. It is a good idea to bypass it using the `{modules:` directive in the scope above. See for example [here](https://erratique.ch/software/bytesrw/doc/#bytesrw) or [there](https://erratique.ch/software/webs/doc/#webs).

**Comment by mjambon** [comment]
> The day you are inflicted a seemingly random grab bag of inlines depending on the API docs you peruse because every API author was allowed to decide the matter based on their taste, your opinions on the matter could perhaps change.

I suggest the following guidelines:
- offer sane defaults
- offer ways to override them

This results in uniformity with exceptions when they're justified.

**Comment by dbuenzli** [comment]
This was already mentioned more than once: everyone's going to have a different view on what is "justified". 

Also by side effect you break a lot of other expectations like a predictable URL structure (a.k.a. hackable URIs).
